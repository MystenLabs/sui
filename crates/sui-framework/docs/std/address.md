---
title: Module `std::address`
---

Provides a way to get address length since it's a
platform-specific parameter.


-  [Function `length`](#std_address_length)


<pre><code></code></pre>



<a name="std_address_length"></a>

## Function `length`

Should be converted to a native function.
Current implementation only works for Sui.


<pre><code><b>public</b> <b>fun</b> <a href="../std/address.md#std_address_length">length</a>(): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/address.md#std_address_length">length</a>(): <a href="../std/u64.md#std_u64">u64</a> {
    32
}
</code></pre>



</details>
