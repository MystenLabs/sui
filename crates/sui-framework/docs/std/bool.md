---
title: Module `std::bool`
---





<pre><code></code></pre>
