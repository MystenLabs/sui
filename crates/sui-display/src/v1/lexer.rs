// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::fmt;

/// Lexer for Display V1 format strings. Format strings are a mix of text and expressions.
/// Expressions are enclosed in curly braces and may contain identifiers separated by dots (a path
/// of field accesses).
#[derive(Debug)]
pub(crate) struct Lexer<'s> {
    /// Remaining input to be tokenized.
    src: &'s str,

    /// The number of bytes tokenized so far.
    off: usize,

    /// Whether the lexer is currently inside a text strand or an expression strand.
    mode: Mode,
}

#[derive(Debug)]
enum Mode {
    Text,
    Expr,
}

/// A lexeme is a token along with its offset in the source string, and the slice of source string
/// that it originated from.
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub(crate) struct Lexeme<'s>(pub Token, pub usize, pub &'s str);

/// Like [Lexeme] but owns the slice of source string. Useful for capturing context in an error
/// message.
#[derive(Debug)]
pub(crate) struct OwnedLexeme(Token, usize, String);

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub(crate) enum Token {
    /// '.'
    Dot,
    /// '\X' where X is any byte.
    Escaped,
    /// A potential field identifier
    Ident,
    /// '{'
    LCurl,
    /// '}'
    RCurl,
    /// A strand of text.
    Text,
}

#[derive(Debug)]
pub(crate) struct TokenSet<'t>(pub &'t [Token]);

impl<'s> Lexer<'s> {
    pub(crate) fn new(src: &'s str) -> Self {
        Self {
            src,
            off: 0,
            mode: Mode::Text,
        }
    }

    /// Assuming the lexer is in text mode, return the next text token.
    fn next_text_token(&mut self) -> Option<Lexeme<'s>> {
        let bytes = self.src.as_bytes();

        use Token as T;
        Some(match bytes.first()? {
            b'\\' if bytes.len() > 1 => {
                self.take(T::Escaped, 1); // discard the backslash
                self.take(T::Escaped, 1)
            }
            b'\\' => self.take(T::Text, 1),
            b'{' => {
                self.mode = Mode::Expr;
                self.take(T::LCurl, 1)
            }
            // This is not a valid token within text, but recognise it so that the parser can
            // produce a better error message.
            b'}' => self.take(T::RCurl, 1),
            _ => self.take_until(T::Text, |c| ['\\', '{', '}'].contains(&c)),
        })
    }

    /// Assuming the lexer is in expression mode, return the next expression token.
    fn next_expr_token(&mut self) -> Option<Lexeme<'s>> {
        self.skip_whitespace();

        use Token as T;
        Some(match self.src.as_bytes().first()? {
            // { is not a valid token within an expression, but recognise it so that the parser can
            // produce a better error message.
            b'{' => self.take(T::LCurl, 1),
            b'}' => {
                self.mode = Mode::Text;
                self.take(T::RCurl, 1)
            }
            b'.' => self.take(T::Dot, 1),
            // The lexer takes a very liberal definition of "identifier", the parser will check
            // whether the identifier is actually valid.
            _ => self.take_until(T::Ident, |c| {
                c.is_whitespace() || c == '.' || c == '{' || c == '}'
            }),
        })
    }

    fn skip_whitespace(&mut self) {
        self.take_until(Token::Text, |c: char| !c.is_whitespace());
    }

    /// Take a prefix of bytes from `self.src` until a byte satisfying pattern `p` is found, and
    /// return it as a lexeme of type `t`. If no such byte is found, take the entire remainder of
    /// the source string.
    fn take_until(&mut self, t: Token, p: impl FnMut(char) -> bool) -> Lexeme<'s> {
        self.take(t, self.src.find(p).unwrap_or(self.src.len()))
    }

    /// Take `n` bytes from the beginning of `self.src` and return them as a lexeme of type `t`.
    ///
    /// ## Safety
    ///
    /// This function assumes that there are at least `n` bytes left in `self.src`, and will panic
    /// if that is not the case.
    fn take(&mut self, t: Token, n: usize) -> Lexeme<'s> {
        let start = self.off;
        let slice = &self.src[..n];
        self.src = &self.src[n..];
        self.off += n;

        Lexeme(t, start, slice)
    }
}

impl Lexeme<'_> {
    /// Return the lexeme as an owned lexeme, with the slice of source string copied.
    pub(crate) fn detach(&self) -> OwnedLexeme {
        OwnedLexeme(self.0, self.1, self.2.to_owned())
    }
}

impl<'s> Iterator for Lexer<'s> {
    type Item = Lexeme<'s>;

    fn next(&mut self) -> Option<Self::Item> {
        use Mode as M;
        match self.mode {
            M::Text => self.next_text_token(),
            M::Expr => self.next_expr_token(),
        }
    }
}

impl fmt::Display for OwnedLexeme {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use OwnedLexeme as L;
        use Token as T;
        match self {
            L(T::Text, _, s) => write!(f, "text {s:?}"),
            L(T::Escaped, _, s) => write!(f, "escaped character '\\{s}'"),
            L(T::LCurl, _, _) => write!(f, "'{{'"),
            L(T::RCurl, _, _) => write!(f, "'}}'"),
            L(T::Ident, _, s) => write!(f, "identifier {s:?}"),
            L(T::Dot, _, _) => write!(f, "'.'"),
        }?;

        write!(f, " at offset {}", self.1)
    }
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use Token as T;
        match self {
            T::Text => write!(f, "text"),
            T::Escaped => write!(f, "an escaped character"),
            T::LCurl => write!(f, "'{{'"),
            T::RCurl => write!(f, "'}}'"),
            T::Ident => write!(f, "an identifier"),
            T::Dot => write!(f, "'.'"),
        }
    }
}

impl fmt::Display for TokenSet<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let TokenSet(tokens) = self;

        if tokens.is_empty() {
            return write!(f, "nothing");
        }

        let (head, [tail]) = tokens.split_at(tokens.len() - 1) else {
            unreachable!("tail contains exactly one token");
        };

        if head.is_empty() {
            return write!(f, "{tail}");
        }

        let mut prefix = "one of ";
        for token in head {
            write!(f, "{prefix}{token}")?;
            prefix = ", ";
        }

        write!(f, ", or {tail}")?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use Lexeme as L;
    use Token as T;

    /// Simple test for a raw literal string.
    #[test]
    fn test_all_text() {
        let lexer = Lexer::new("foo bar");
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(lexemes, vec![L(T::Text, 0, "foo bar")]);
    }

    /// Escape sequences are all text, but
    #[test]
    fn test_escapes() {
        let lexer = Lexer::new(r#"foo \b\{ar\}"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Text, 0, "foo "),
                L(T::Escaped, 5, "b"),
                L(T::Escaped, 7, "{"),
                L(T::Text, 8, "ar"),
                L(T::Escaped, 11, "}"),
            ]
        );
    }

    /// If the last character is a backslash, then treat it as just a backslash, not the start of
    /// an escape sequence.
    #[test]
    fn test_trailing_escape() {
        let lexer = Lexer::new(r#"foo bar\"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(lexemes, vec![L(T::Text, 0, "foo bar"), L(T::Text, 7, "\\")],);
    }

    /// Text inside curly braces is tokenized as if it's an expression.
    #[test]
    fn test_expressions() {
        let lexer = Lexer::new(r#"foo {bar}"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Text, 0, "foo "),
                L(T::LCurl, 4, "{"),
                L(T::Ident, 5, "bar"),
                L(T::RCurl, 8, "}"),
            ],
        );
    }

    /// Expressions are tokenized to ignore whitespace.
    #[test]
    fn test_expression_whitespace() {
        let lexer = Lexer::new(r#"foo {  bar   }"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Text, 0, "foo "),
                L(T::LCurl, 4, "{"),
                L(T::Ident, 7, "bar"),
                L(T::RCurl, 13, "}"),
            ],
        );
    }

    /// Field names are separated by dots in an expression.
    #[test]
    fn test_expression_dots() {
        let lexer = Lexer::new(r#"foo {bar. baz  . qux}"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Text, 0, "foo "),
                L(T::LCurl, 4, "{"),
                L(T::Ident, 5, "bar"),
                L(T::Dot, 8, "."),
                L(T::Ident, 10, "baz"),
                L(T::Dot, 15, "."),
                L(T::Ident, 17, "qux"),
                L(T::RCurl, 20, "}"),
            ],
        );
    }

    /// Multiple expressions test switching and back and forth between lexer modes.
    #[test]
    fn test_multiple_expressions() {
        let lexer = Lexer::new(r#"foo {bar.baz} qux {quy.quz}"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Text, 0, "foo "),
                L(T::LCurl, 4, "{"),
                L(T::Ident, 5, "bar"),
                L(T::Dot, 8, "."),
                L(T::Ident, 9, "baz"),
                L(T::RCurl, 12, "}"),
                L(T::Text, 13, " qux "),
                L(T::LCurl, 18, "{"),
                L(T::Ident, 19, "quy"),
                L(T::Dot, 22, "."),
                L(T::Ident, 23, "quz"),
                L(T::RCurl, 26, "}"),
            ],
        );
    }

    /// Left curlies are not valid inside expressions and right curlies are not valid inside text
    /// strands, but we still tokenize them so that we can detect them during parsing and return
    /// ane error message.
    #[test]
    fn test_misplaced_curlies() {
        let lexer = Lexer::new(r#"foo}{bar{}}"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Text, 0, "foo"),
                L(T::RCurl, 3, "}"),
                L(T::LCurl, 4, "{"),
                L(T::Ident, 5, "bar"),
                L(T::LCurl, 8, "{"),
                L(T::RCurl, 9, "}"),
                L(T::RCurl, 10, "}"),
            ],
        );
    }

    /// The lexer is very permissive about what it considers an identifier, this allows it to
    /// gather more context without failing, while the parser will check that the identifier is
    /// valid.
    #[test]
    fn test_strange_identifiers() {
        let lexer = Lexer::new(r#"{ not-really . an! . ident#fier? }"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::LCurl, 0, "{"),
                L(T::Ident, 2, "not-really"),
                L(T::Dot, 13, "."),
                L(T::Ident, 15, "an!"),
                L(T::Dot, 19, "."),
                L(T::Ident, 21, "ident#fier?"),
                L(T::RCurl, 33, "}"),
            ],
        );
    }

    /// The lexer should correctly identify backslashes that signify escapes vs backslashes that
    /// are literal.
    #[test]
    fn test_escape_chain() {
        let lexer = Lexer::new(r#"\\\\\\\\\"#);
        let lexemes: Vec<_> = lexer.collect();
        assert_eq!(
            lexemes,
            vec![
                L(T::Escaped, 1, r#"\"#),
                L(T::Escaped, 3, r#"\"#),
                L(T::Escaped, 5, r#"\"#),
                L(T::Escaped, 7, r#"\"#),
                L(T::Text, 8, r#"\"#),
            ],
        );
    }
}
