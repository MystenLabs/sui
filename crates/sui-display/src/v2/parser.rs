// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0
#![allow(dead_code)]

use std::{borrow::Cow, iter::Peekable};

use move_core_types::{
    account_address::AccountAddress,
    language_storage::{StructTag, TypeTag},
    u256::U256,
};

use super::lexer::{Lexeme as L, Lexer, OwnedLexeme, Token as T, TokenSet};

/// A single Display string template is a sequence of strands.
#[derive(Debug)]
pub enum Strand<'s> {
    /// Text strands are ported literally to the output.
    Text(Cow<'s, str>),

    /// Expr strands are interpreted to some value whose string representation is included in the
    /// output. They are surrounded by curly braces in the syntax, to differentiate them from text.
    Expr(Expr<'s>),
}

/// Expressions are composed of a number of alternates and an optional transform. During
/// evaluation, each alternate is evaluated in turn until the first one succeeds, and if a
/// transform is provided, it is applied to the result to convert it to a string.
#[derive(Debug)]
pub struct Expr<'s> {
    alternates: Vec<Chain<'s>>,
    transform: Option<&'s str>,
}

/// Chains are a sequence of nested field accesses.
#[derive(Debug)]
pub struct Chain<'s> {
    /// An optional root expression. If not provided, the object being displayed is the root.
    root: Option<Literal<'s>>,

    /// A sequence of field accessors that go successively deeper into the object.
    accessors: Vec<Accessor<'s>>,
}

/// Different ways to nest deeply into an object.
#[derive(Debug)]
pub enum Accessor<'s> {
    /// Access a named field.
    Field(&'s str),

    /// Index into a vector, VecMap, or dynamic field.
    Index(Chain<'s>),

    /// Index into a dynamic object field.
    IIndex(Chain<'s>),
}

/// Literal forms are elements whose syntax determines their (outer) type.
#[derive(Debug)]
pub enum Literal<'s> {
    // Primitives
    Address(AccountAddress),
    Bool(bool),
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    U256(U256),

    // Aggregates
    ByteArray(Vec<u8>),
    String(Cow<'s, str>),
    Vector(Box<Vector<'s>>),

    // Datatypes
    Struct(Box<Struct<'s>>),
    Enum(Box<Enum<'s>>),
}

/// Contents of a vector literal.
#[derive(Debug)]
pub struct Vector<'s> {
    /// Element type, optional for non-empty vectors.
    type_: Option<TypeTag>,
    elements: Vec<Chain<'s>>,
}

/// Contents of a struct literal.
#[derive(Debug)]
pub struct Struct<'s> {
    type_: StructTag,
    fields: Fields<'s>,
}

/// Contents of an enum literal.
#[derive(Debug)]
pub struct Enum<'s> {
    type_: StructTag,
    variant_name: Option<&'s str>,
    variant_index: u16,
    fields: Fields<'s>,
}

#[derive(Debug)]
pub enum Fields<'s> {
    Positional(Vec<Chain<'s>>),
    Named(Vec<(&'s str, Chain<'s>)>),
}

pub(crate) struct Parser<'s> {
    lexer: Peekable<Lexer<'s>>,
}

#[derive(thiserror::Error, Debug)]
pub enum Error {
    #[error("Unexpected end-of-string, expected {expect}")]
    UnexpectedEos { expect: TokenSet<'static> },

    #[error("Unexpected {actual}, expected {expect}")]
    UnexpectedToken {
        actual: OwnedLexeme,
        expect: TokenSet<'static>,
    },
}

/// Pattern match on the next token in the lexer, without consuming it. Returns an error if there
/// is no next token, or if the next token doesn't match any of the provided patterns. The error
/// enumerates all the tokens that were expected.
macro_rules! match_token {
    ($lexer:expr; $(L($($pat:path)|+, $off:pat, $slice:pat) => $expr:expr),+ $(,)?) => {{
        const EXPECTED: TokenSet = TokenSet(&[$($($pat),+),+]);

        match $lexer.peek().ok_or_else(|| Error::UnexpectedEos { expect: EXPECTED })? {
            $(&L($($pat)|+, $off, $slice) => $expr,)+
            &actual => return Err(Error::UnexpectedToken {
                actual: actual.detach(),
                expect: EXPECTED,
            }),
        }
    }};
}

// Expression variant of `match_token!`, which evaluates to `Some(...)` if an arm matches,
// consuming the next token, or `None` without consuming if there is no next token, or it doesn't
// match any of the provided patterns.
macro_rules! match_token_opt {
    ($lexer:expr; $(L($($pat:path)|+, $off:pat, $slice:pat) => $expr:expr),+ $(,)?) => {{
        match $lexer.peek() {
            $(Some(&L($($pat)|+, $off, $slice)) => Some($expr),)+
            Some(_) | None => None
        }
    }};
}

/// Recursive descent parser for Display V2 format strings, parsing the following grammar:
///
///   format   ::= strand*
///
///   strand   ::= text | expr
///
///   text     ::= part+
///
///   part     ::= TEXT | '{{' | '}}'
///
///   expr     ::= '{' chain ('|' chain)* (':' IDENT)? '}'
///
///   chain    ::= (literal | IDENT) accessor*
///
///   accessor ::= '.' IDENT
///              | '[' chain ']'
///              | '[' '[' chain ']' ']'
///
///   literal  ::= address | bool | number | string | vector | struct | enum
///
///   address  ::= '@' NUM_HEX
///
///   bool     ::= 'true' | 'false'
///
///   number   ::= (NUM_DEC | NUM_HEX) numeric?
///
///   string   ::= ('b' | 'x')? STRING
///
///   vector   ::= 'vector'  '<' type '>'
///              | 'vector' ('<' type '>')? '[' chain (',' chain)* ','? ']'
///
///   struct   ::= datatype fields
///
///   enum     ::= datatype '::' (IDENT '#')? NUM_DEC fields
///
///   fields   ::= '(' chain (',' chain)* ','? ')'
///              | '{' named (',' named)* ','? '}'
///
///   named    ::= IDENT ':' chain
///
///   type     ::= 'address' | 'bool' | | 'vector' '<' type '>' |  numtype | datatype
///
///   datatype ::= NUM_HEX '::' IDENT ('<' type (',' type)* ','? '>')?
///
///   numeric  ::= 'u8' | 'u16' | 'u32' | 'u64' | 'u128' | 'u256'
///
impl<'s> Parser<'s> {
    /// Construct a new parser, consuming input from the `src` string.
    pub(crate) fn new(src: &'s str) -> Self {
        Self {
            lexer: Lexer::new(src).peekable(),
        }
    }

    /// Entrypoint into the parser, parsing the root non-terminal -- `format`. Consumes all the
    /// remaining input in the parser and the parser itself.
    pub(crate) fn parse_format(mut self) -> Result<Vec<Strand<'s>>, Error> {
        let mut strands = vec![];
        while self.lexer.peek().is_some() {
            strands.push(self.parse_strand()?);
        }

        Ok(strands)
    }

    fn parse_strand(&mut self) -> Result<Strand<'s>, Error> {
        Ok(match_token! { self.lexer;
            L(T::Text | T::LLBrace | T::RRBrace, _, _) => Strand::Text(self.parse_text()?),
            L(T::LBrace, _, _) => Strand::Expr(self.parse_expr()?),
        })
    }

    fn parse_text(&mut self) -> Result<Cow<'s, str>, Error> {
        let mut text = self.parse_part()?;
        while let Some(L(T::Text | T::LLBrace | T::RRBrace, _, _)) = self.lexer.peek() {
            text += self.parse_part()?;
        }

        Ok(text)
    }

    fn parse_part(&mut self) -> Result<Cow<'s, str>, Error> {
        Ok(match_token! { self.lexer;
            L(T::Text | T::LLBrace | T::RRBrace, _, slice) => {
                self.lexer.next();
                Cow::Borrowed(slice)
            }
        })
    }

    fn parse_expr(&mut self) -> Result<Expr<'s>, Error> {
        match_token! { self.lexer; L(T::LBrace, _, _) => self.lexer.next() };
        let mut alternates = vec![self.parse_chain()?];
        let mut transform = None;

        loop {
            self.eat_whitespace();
            match_token! { self.lexer;
                L(T::RBrace, _, _) => {
                    self.lexer.next();
                    break;
                },

                L(T::Colon, _, _) => {
                    self.lexer.next();
                    self.eat_whitespace();
                    match_token! { self.lexer; L(T::Ident, _, t) => {
                        self.lexer.next();
                        transform = Some(t);
                    }};
                    self.eat_whitespace();
                    match_token! { self.lexer; L(T::RBrace, _, _) => {
                        self.lexer.next()
                    }};
                    break;
                },

                L(T::Pipe, _, _) => {
                    self.lexer.next();
                    alternates.push(self.parse_chain()?);
                }
            }
        }

        Ok(Expr {
            alternates,
            transform,
        })
    }

    fn parse_chain(&mut self) -> Result<Chain<'s>, Error> {
        let mut accessors = vec![];
        let root = self.try_parse_literal()?;

        // If there is no root literal, the chain must start with an identifier, representing a
        // field on the object being displayed.
        if root.is_none() {
            accessors.push(match_token! { self.lexer; L(T::Ident, _, i) => {
                self.lexer.next();
                Accessor::Field(i)
            }})
        }

        while let Some(accessor) = self.try_parse_accessor()? {
            accessors.push(accessor);
        }

        Ok(Chain { root, accessors })
    }

    fn try_parse_literal(&mut self) -> Result<Option<Literal<'s>>, Error> {
        self.eat_whitespace();
        Ok(match_token_opt! { self.lexer;
            L(T::Ident, _, "true") => {
                self.lexer.next();
                Literal::Bool(true)
            },
        })
    }

    fn try_parse_accessor(&mut self) -> Result<Option<Accessor<'s>>, Error> {
        self.eat_whitespace();
        Ok(match_token_opt! { self.lexer;
            L(T::Dot, _, _) => {
                self.lexer.next();
                self.eat_whitespace();
                match_token! { self.lexer; L(T::Ident, _, f) => {
                    self.lexer.next();
                    Accessor::Field(f)
                }}
            },

            L(T::LBracket, _, _) => {
                self.lexer.next();
                if matches!(self.lexer.peek(), Some(L(T::LBracket, _, _))) {
                    self.lexer.next();
                    let chain = self.parse_chain()?;
                    self.eat_whitespace();
                    match_token! { self.lexer; L(T::RBracket, _, _) => self.lexer.next() };
                    match_token! { self.lexer; L(T::RBracket, _, _) => self.lexer.next() };
                    Accessor::IIndex(chain)
                } else {
                    let chain = self.parse_chain()?;
                    self.eat_whitespace();
                    match_token! { self.lexer; L(T::RBracket, _, _) => self.lexer.next() };
                    Accessor::Index(chain)
                }
            },
        })
    }

    fn eat_whitespace(&mut self) {
        // The lexer merges together consecutive whitespace tokens, so if one is found, there is no
        // need to check for more.
        if let Some(L(T::Whitespace, _, _)) = self.lexer.peek() {
            self.lexer.next();
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;

    fn strands(src: &str) -> String {
        let strands = match Parser::new(src).parse_format() {
            Ok(strands) => strands,
            Err(e) => return format!("Error: {e}"),
        };

        let mut output = String::new();
        for strand in strands {
            output += &format!("{strand:#?}\n");
        }

        output
    }

    #[test]
    fn test_all_text() {
        assert_snapshot!(strands(r#"foo bar"#), @r###"
        Text(
            "foo bar",
        )
        "###);
    }

    #[test]
    fn test_field_expr() {
        assert_snapshot!(strands(r#"{foo}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_literal_expr() {
        assert_snapshot!(strands(r#"{true}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: Some(
                            Bool(
                                true,
                            ),
                        ),
                        accessors: [],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_nested_field_expr() {
        assert_snapshot!(strands(r#"{foo.bar.baz}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                            Field(
                                "bar",
                            ),
                            Field(
                                "baz",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_text_with_escapes() {
        assert_snapshot!(strands(r#"foo {{bar}} baz"#), @r###"
        Text(
            "foo {bar} baz",
        )
        "###);
    }

    #[test]
    fn test_back_to_back_exprs() {
        assert_snapshot!(strands(r#"{foo . bar}{baz.qux}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                            Field(
                                "bar",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "baz",
                            ),
                            Field(
                                "qux",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_triple_curlies() {
        assert_snapshot!(strands(r#"foo {{{bar} {baz}}}"#), @r###"
        Text(
            "foo {",
        )
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "bar",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        Text(
            " ",
        )
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "baz",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        Text(
            "}",
        )
        "###);
    }

    #[test]
    fn test_alternates() {
        assert_snapshot!(strands(r#"{foo | bar | baz}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                        ],
                    },
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "bar",
                            ),
                        ],
                    },
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "baz",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_alternates_with_transform() {
        assert_snapshot!(strands(r#"{foo | bar | baz :base64}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                        ],
                    },
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "bar",
                            ),
                        ],
                    },
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "baz",
                            ),
                        ],
                    },
                ],
                transform: Some(
                    "base64",
                ),
            },
        )
        "###);
    }

    #[test]
    fn test_expr_with_transform() {
        assert_snapshot!(strands(r#"{foo.bar.baz:url}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                            Field(
                                "bar",
                            ),
                            Field(
                                "baz",
                            ),
                        ],
                    },
                ],
                transform: Some(
                    "url",
                ),
            },
        )
        "###);
    }

    #[test]
    fn test_index_chain() {
        assert_snapshot!(strands(r#"{foo[bar][[baz]].qux[quy]}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: None,
                        accessors: [
                            Field(
                                "foo",
                            ),
                            Index(
                                Chain {
                                    root: None,
                                    accessors: [
                                        Field(
                                            "bar",
                                        ),
                                    ],
                                },
                            ),
                            IIndex(
                                Chain {
                                    root: None,
                                    accessors: [
                                        Field(
                                            "baz",
                                        ),
                                    ],
                                },
                            ),
                            Field(
                                "qux",
                            ),
                            Index(
                                Chain {
                                    root: None,
                                    accessors: [
                                        Field(
                                            "quy",
                                        ),
                                    ],
                                },
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_index_with_root() {
        assert_snapshot!(strands(r#"{true[[foo]][bar].baz}"#), @r###"
        Expr(
            Expr {
                alternates: [
                    Chain {
                        root: Some(
                            Bool(
                                true,
                            ),
                        ),
                        accessors: [
                            IIndex(
                                Chain {
                                    root: None,
                                    accessors: [
                                        Field(
                                            "foo",
                                        ),
                                    ],
                                },
                            ),
                            Index(
                                Chain {
                                    root: None,
                                    accessors: [
                                        Field(
                                            "bar",
                                        ),
                                    ],
                                },
                            ),
                            Field(
                                "baz",
                            ),
                        ],
                    },
                ],
                transform: None,
            },
        )
        "###);
    }

    #[test]
    fn test_unbalanced_curlies() {
        assert_snapshot!(strands(r#"{foo"#), @"Error: Unexpected end-of-string, expected one of '}', ':', or '|'");
    }

    #[test]
    fn test_missing_field_identifier() {
        assert_snapshot!(strands(r#"{foo..bar}"#), @"Error: Unexpected '.' at offset 5, expected an identifier");
    }

    #[test]
    fn test_unbalanced_index() {
        assert_snapshot!(strands(r#"{foo[bar}"#), @"Error: Unexpected '}' at offset 8, expected ']'");
    }

    #[test]
    fn test_spaced_out_left_double_index() {
        assert_snapshot!(strands(r#"{foo[ [bar]]}"#), @"Error: Unexpected '[' at offset 6, expected an identifier");
    }

    #[test]
    fn test_spaced_out_right_double_index() {
        assert_snapshot!(strands(r#"{foo[[bar] ]}"#), @"Error: Unexpected whitespace at offset 10, expected ']'");
    }

    #[test]
    fn test_unbalanced_double_index() {
        assert_snapshot!(strands(r#"{foo[[bar}"#), @"Error: Unexpected '}' at offset 9, expected ']'");
    }

    #[test]
    fn test_triple_index() {
        assert_snapshot!(strands(r#"{foo[[[bar]]]}"#), @"Error: Unexpected '[' at offset 6, expected an identifier");
    }

    #[test]
    fn test_unexpected_characters() {
        assert_snapshot!(strands(r#"anything goes {? % ! 🔥}"#), @r###"Error: Unexpected "?" at offset 15, expected an identifier"###);
    }
}
