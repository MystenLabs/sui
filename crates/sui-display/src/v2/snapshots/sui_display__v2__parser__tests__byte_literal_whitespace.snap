---
source: crates/sui-display/src/v2/parser.rs
expression: "strands(r#\"{b 'foo'}\"#)"
---
Error: Unexpected whitespace followed by string "foo" at offset 4, expected one of ':', '|', or '}'
